core/src/main/java/com/scalar/db/util/groupcommit/GroupManager.java

```diff
@@ -0,0 +1,220 @@
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.errorprone.annotations.ThreadSafe;
+import com.google.errorprone.annotations.concurrent.LazyInit;
+import com.scalar.db.util.groupcommit.KeyManipulator.Keys;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.locks.StampedLock;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@ThreadSafe
+class GroupManager<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> {
+  private static final Logger logger = LoggerFactory.getLogger(GroupManager.class);
+
+  // Groups
+  @Nullable private NormalGroup<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> currentGroup;
+  // Note: Using ConcurrentHashMap results in less performance.
+  private final Map<PARENT_KEY, NormalGroup<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V>>
+      normalGroupMap = new HashMap<>();
+  private final Map<FULL_KEY, DelayedGroup<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V>>
+      delayedGroupMap = new HashMap<>();
+  // Only this class uses this type of lock since the class can be heavy hotspot and StampedLock has
+  // basically better performance than `synchronized` keyword.
+  private final StampedLock lock = new StampedLock();
+
+  // Background workers
+  @LazyInit private GroupCloseWorker<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> groupCloseWorker;
+
+  @LazyInit
+  private GroupCleanupWorker<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> groupCleanupWorker;
+
+  // Custom operations injected by the client
+  private final KeyManipulator<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY> keyManipulator;
+  @LazyInit private Emittable<EMIT_KEY, V> emitter;
+
+  private final long groupCloseTimeoutMillis;
+  private final long delayedSlotMoveTimeoutMillis;
+  private final int slotCapacity;
+
+  GroupManager(
+      String label,
+      GroupCommitConfig config,
+      KeyManipulator<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY> keyManipulator) {
+    this.keyManipulator = keyManipulator;
+    this.groupCloseTimeoutMillis = config.groupCloseTimeoutMillis();
+    this.delayedSlotMoveTimeoutMillis = config.delayedSlotMoveTimeoutMillis();
+    this.slotCapacity = config.slotCapacity();
+  }
+
+  void setGroupCloseWorker(
+      GroupCloseWorker<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> groupCloseWorker) {
+    this.groupCloseWorker = groupCloseWorker;
+  }
+
+  void setGroupCleanupWorker(
+      GroupCleanupWorker<PARENT_KEY, CHILD_KEY, FULL_KEY, EMIT_KEY, V> groupCleanupWorker) {
+    this.groupCleanupWorker = groupCleanupWorker;
+  }
+
+  // Reserves a new slot in the current NormalGroup. A new NormalGroup will be created and
+  // registered to `normalGroupMap` if the current NormalGroup is already closed.
+  //
+  // If it returns null, the Group is already closed and a retry is needed.
+  @Nullable
+  FULL_KEY reserveNewSlot(CHILD_KEY childKey) {
+    long stamp = lock.writeLock();
```

Comment:
I might miss something, but since we have only one active group at a time, I'm wondering how much it badly affects performance if there are many concurrent transactions.
We could use multiple active groups to mitigate the potential issue, but it might complicate the mechanism.
Any thoughts on this?

Reply:
Good point. I also thought and tried multiple current (== active) groups before. But, the performance with multiple active groups didn't have a good performance. I think how fast to fill up a group with slots (and values) is a very important factor to move it to the next state (i.e., ready to commit) in the group commit and using multiple active groups doesn't fit.

I benchmarked this `util/groupcommit` pacakge itself and the average throughput was about 65K TPS without any wait while the throughput with some delays to emulate actual latencies of underlying database had far low throughputs. It means current single active group won't be a bottleneck in usual usages of ScalarDB, I think.

Reply:
Thank you for the detailed explanation!
It makes sense. Let's leave it as is.